vertx/src/main/java/feign/VertxFeign.java

```diff
@@ -0,0 +1,476 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package feign;
+
+import static feign.Util.checkNotNull;
+import static feign.Util.isDefault;
+
+import feign.InvocationHandlerFactory.MethodHandler;
+import feign.codec.Decoder;
+import feign.codec.Encoder;
+import feign.codec.ErrorDecoder;
+import feign.querymap.FieldQueryMapEncoder;
+import feign.vertx.VertxDelegatingContract;
+import feign.vertx.VertxHttpClient;
+import io.vertx.core.Vertx;
+import io.vertx.core.http.HttpClient;
+import io.vertx.core.http.HttpClientOptions;
+import io.vertx.core.http.HttpClientRequest;
+import java.lang.reflect.InvocationHandler;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.UnaryOperator;
+
+/**
+ * Allows Feign interfaces to return Vert.x {@link io.vertx.core.Future Future}s.
+ *
+ * @author Alexei KLENIN
+ * @author Gordon McKinney
+ */
+public final class VertxFeign extends Feign {
+  private final ParseHandlersByName targetToHandlersByName;
+  private final InvocationHandlerFactory factory;
+
+  private VertxFeign(
+      final ParseHandlersByName targetToHandlersByName, final InvocationHandlerFactory factory) {
+    this.targetToHandlersByName = targetToHandlersByName;
+    this.factory = factory;
+  }
+
+  public static Builder builder() {
+    return new Builder();
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public <T> T newInstance(final Target<T> target) {
+    checkNotNull(target, "Argument target must be not null");
+
+    final Map<String, MethodHandler> nameToHandler = targetToHandlersByName.apply(target);
+    final Map<Method, MethodHandler> methodToHandler = new HashMap<>();
+    final List<DefaultMethodHandler> defaultMethodHandlers = new ArrayList<>();
+
+    for (final Method method : target.type().getMethods()) {
+      if (isDefault(method)) {
+        final DefaultMethodHandler handler = new DefaultMethodHandler(method);
+        defaultMethodHandlers.add(handler);
+        methodToHandler.put(method, handler);
+      } else {
+        methodToHandler.put(method, nameToHandler.get(Feign.configKey(target.type(), method)));
+      }
+    }
+
+    final InvocationHandler handler = factory.create(target, methodToHandler);
+    final T proxy =
+        (T)
+            Proxy.newProxyInstance(
+                target.type().getClassLoader(), new Class<?>[] {target.type()}, handler);
+
+    for (final DefaultMethodHandler defaultMethodHandler : defaultMethodHandlers) {
+      defaultMethodHandler.bindTo(proxy);
+    }
+
+    return proxy;
+  }
+
+  /** VertxFeign builder. */
+  public static final class Builder extends Feign.Builder {
```

Comment:
You might want to take a look on this one:
https://github.com/OpenFeign/feign/blob/master/core/src/main/java/feign/AsyncFeign.java#L69

Reply:
Checked it already. `AsyncFeign` is unfit because it's based on `CompletableFuture`s from JDK. As I told before, Verx uses its own `Future` implementation, which should be supported both internally and used in contracts. My project is more similar to `feign-rective` in that sense.